### PR TITLE
i#6299: Collapse redundant entries in schedule files

### LINKED ISSUE #6299: Collapse redundant entries in drmemtrace schedule files
The serial and per-cpu schedule files store by raw2trace have an entry for every timestamp, even for consecutive entries with the same thread, which are redundant as there is no switch.  With new timestamps being added for #6289 there will be even more of these entries.  We should collapse them in raw2trace.  This also requires updating invariant_checker to collapse them as well for its checks.

### PR DESCRIPTION
**Pull request by derekbruening**
Before, we wrote out a schedule file record for every timestamp.  But for scheduling we only need to know the thread switch points, so consecutive same-thread records are superfluous.  This will only get worse with more timestamps from #6290.  Here we collapse consecutive same-thread entries in both the serial and cpu schedule files, and update the invariant checker to allow either the old or new type of file.

On simple_app, before (2nd column is tid):

```
$ zcat drmemtrace.*.dir/trace/serial* | od -t d8 -A d -w32 | tail
0002464              1484994    13338489228331182        3               112619
0002496              1484994    13338489228331508        3               112619
0002528              1484994    13338489228337025        3               112886
0002560              1484994    13338489228337031        3               112886
0002592              1484994    13338489228360665        3               113878
0002624              1484994    13338489228360692        3               113878
0002656              1484994    13338489228437577        3               116138
0002688              1484994    13338489228437586        3               116138
0002720              1484994    13338489228520213        3               118090
0002752

$ unzip -p drmemtrace.*.dir/trace/cpu*.zip 3 | od -t d8 -A d -w32 | tail
000520              1484994    13338489228331182        3               112619
000540              1484994    13338489228331508        3               112619
000560              1484994    13338489228337025        3               112886
000580              1484994    13338489228337031        3               112886
0005a0              1484994    13338489228360665        3               113878
0005c0              1484994    13338489228360692        3               113878
0005e0              1484994    13338489228437577        3               116138
000600              1484994    13338489228437586        3               116138
000620              1484994    13338489228520213        3               118090
000640
```

After:
```
$ zcat drmemtrace.*.dir/trace/serial* | od -t d8 -A d -w32 | tail
0000000              1873100    13338569448213406       1                    0
0000032

$ unzip -p drmemtrace.*.dir/trace/cpu*.zip 1 | od -t d8 -A x -w32 | tail
000000              1873100    13338569448213406        1                    0
000020
```

Fixes #6299

### CODE REVIEW

**Comment by derekbruening** [comment]
The doxygen download from sourceforge is down (again) -- was always transient in the past.  (Doesn't look like a doxygen version drop as the sourceforge page has a big banner about an outage.). This is why all the Windows runs are failing.  Should be fine for review to go forward.  Will wait to see if doxygen comes back up later today before merging.

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 1]
```diff
@@ -889,24 +895,33 @@ invariant_checker_t::check_schedule_data(per_shard_t *global)
                       << " records from the file and observed " << serial.size()
                       << " transition in the trace\n";
         }
-        report_if_false(global, serial_file.size() == serial.size(),
-                        "Serial schedule entry count does not match trace");
-        for (int i = 0; i < static_cast<int>(serial.size()) &&
+        std::vector<schedule_entry_t> *tomatch;
+        if (serial_file.size() == serial.size())
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:899]: Done.  There is also the comment above too.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:900]: Thanks.

You mean:
```
After i#6299, these files collapse same-thread entries.
```

That doesn't make it clear that we intend to support both formats for the near future.

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:900]: Already augmented that to make it clearer in [e48004d](https://github.com/DynamoRIO/dynamorio/pull/6301/commits/e48004d1b34aacc1453682ee7dbfc28db8b52dc7).

**Pull request merged**